### PR TITLE
JAXArrayContext

### PR DESCRIPTION
**Pull request by kaushikcfd**
TODO:
- <del>[ ] Map loopy IR to python callable consisting of non-stateful array operations.</del>
  - Dropping because translating logic is complicated, most likely unnecessary.
- [x] Test JAXArrayContext
- [x] Needs https://github.com/inducer/pytato/pull/233

### CODE REVIEW

File: arraycontext/impl/jax/__init__.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    A :class:`ArrayContext` that uses :mod:`jax.numpy.DeviceArray` instances
+    for its base array class. Performs all array operations eagerly. See
+    :class:`~arraycontext.PytatoJAXArrayContext` for a lazier version.
+
+    .. note::
+
+        JAX stores a global configuration state in the module
+        :mod:`jax.config`. Callers are expected to maintain those. Most
+        important for scientific computing workloads being
+        ``jax_enable_x64``.
```
Comment (alexfikl) [inline arraycontext/impl/jax/__init__.py:49]: This doesn't seem like it will be relevant for much longer
https://github.com/google/jax/issues/8178

Comment (kaushikcfd) [reply arraycontext/impl/jax/__init__.py:49]: Since that's still not resolved. I'll keep the note at this point. (Thanks!)

Comment (alexfikl) [reply arraycontext/impl/jax/__init__.py:49]: Yeah, definitely. I hoped they'd fix it by now, but it looks like it stalled :cry:

**Review by alexfikl — COMMENTED** [review]
This all looks good to me, so just left a bunch of cleanup comments here and there!

(I didn't look at the `pytato.compile` parts though)

File: arraycontext/impl/pytato/__init__.py [hunk 2]
```diff
@@ -315,14 +386,140 @@ def preprocess_arg(name, arg):
...
-    @property
-    def permits_inplace_modification(self):
-        return False
 
-    @property
-    def supports_nonscalar_broadcasting(self):
-        return True
+class PytatoJAXArrayContext(_BasePytatoArrayContext):
+    """
+    An arraycontext that uses a :mod:`pytato` to represent the thawed state of
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py]: Yep. Got rid of overriding `arctan2` (+others like `equal`, `greater`, ...)

File: arraycontext/impl/pytato/__init__.py [hunk 3]
```diff
@@ -44,14 +46,108 @@
...
+                         "transform_loopy_program. Sub-classes are supposed "
+                         "to implement it.")
+
+    @abc.abstractmethod
+    def tag(self, tags: ToTagSetConvertible, array):
+        pass
+
+    @abc.abstractmethod
+    def tag_axis(self, iaxis, tags: ToTagSetConvertible, array):
+        pass
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py]: Yep, got rid of these.

File: arraycontext/impl/pytato/__init__.py [hunk 4]
```diff
@@ -44,14 +46,108 @@
...
+        :mod:`pytato` DAG that is compiled to a GPU-kernel is
+        passed through this routine.
+
+        :arg dag: An instance of :class:`pytato.DictOfNamedArrays`
+        :returns: A transformed version of *dag*.
+        """
+        return dag
+
+    def transform_loopy_program(self, t_unit):
+        raise ValueError("_BasePytatoArrayContext does not implement "
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py]: Thanks, done!

File: arraycontext/impl/pytato/__init__.py [hunk 5]
```diff
@@ -44,14 +46,108 @@
...
+
+    def _get_fake_numpy_namespace(self):
+        from arraycontext.impl.pytato.fake_numpy import PytatoFakeNumpyNamespace
+        return PytatoFakeNumpyNamespace(self)
+
+    def clone(self):
+        raise NotImplementedError(f".clone() not supported for {type(self)}.")
+
+    def empty(self, shape, dtype):
+        raise ValueError("PytatoPyOpenCLArrayContext does not support empty")
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py:77]: Thanks!

File: arraycontext/impl/jax/fake_numpy.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        return rec_map_array_container(_rec_ravel, a)
+
+    def vdot(self, x, y, dtype=None):
+        from arraycontext import rec_multimap_reduce_array_container
+
+        def _rec_vdot(ary1, ary2):
+            if dtype not in [None, numpy.find_common_type((ary1.dtype,
+                                                           ary2.dtype),
+                                                          ())]:
+                raise NotImplementedError("JAXArrayContext cannot take dtype in"
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/fake_numpy.py]: Thansk!

File: arraycontext/impl/jax/fake_numpy.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    def less(self, x, y):
+        return rec_multimap_array_container(jnp.less, x, y)
+
+    def less_equal(self, x, y):
+        return rec_multimap_array_container(jnp.less_equal, x, y)
+
+    # }}}
+
+    def arctan2(self, y, x):
+        return rec_multimap_array_container(jnp.arctan2, y, x)
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/fake_numpy.py]: Yep made it go through that. Thanks!

File: arraycontext/impl/jax/fake_numpy.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            a)
+
+    def transpose(self, a, axes=None):
+        return rec_multimap_array_container(jnp.transpose, a, axes)
+
+    def concatenate(self, arrays, axis=0):
+        return rec_multimap_array_container(jnp.concatenate, arrays, axis)
+
+    def ones_like(self, ary):
+        return jnp.ones_like(ary)
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/fake_numpy.py]: Done, thanks

File: arraycontext/impl/jax/fake_numpy.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        return JAXFakeNumpyLinalgNamespace(self._array_context)
+
+    _jax_funcs = {"abs", "sin", "cos", "tan", "arcsin", "arccos",
+                  "arctan", "sinh", "cosh", "tanh", "exp", "log", "log10",
+                  "isnan", "sqrt", "exp", "conj"}
+
+    def __getattr__(self, name):
+        if name in self._jax_funcs:
+            from functools import partial
+            return partial(rec_map_array_container, getattr(jnp, name))
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/fake_numpy.py]: Thanks, done!

File: arraycontext/__init__.py [hunk 10]
```diff
@@ -64,9 +64,13 @@
         outer)
 
 from .impl.pyopencl import PyOpenCLArrayContext
-from .impl.pytato import PytatoPyOpenCLArrayContext
+from .impl.pytato import (PytatoPyOpenCLArrayContext,
+                          PytatoJAXArrayContext,
+                          _BasePytatoArrayContext)
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/__init__.py]: No reason, got rid of it. Thanks!

File: arraycontext/impl/jax/__init__.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+from typing import Union, Callable, Any
+from pytools.tag import ToTagSetConvertible
+from arraycontext.context import ArrayContext, _ScalarLike
+
+
+class EagerJAXArrayContext(ArrayContext):
+    """
+    A :class:`ArrayContext` that uses :mod:`jax.numpy.DeviceArray` instances
+    for its base array class. Performs all array operations eagerly. See
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/__init__.py]: Thanks!

File: arraycontext/impl/jax/__init__.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
+"""
+.. currentmodule:: arraycontext
+.. autoclass:: EagerJAXArrayContext
+.. autoclass:: JITCompilingJAXArrayContext
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/__init__.py]: Thanks. Realized that this file wasn't being traversed by sphinx, added it here: https://github.com/inducer/arraycontext/blob/main/doc/array_context.rst.

File: arraycontext/impl/jax/fake_numpy.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+class JAXFakeNumpyNamespace(BaseFakeNumpyNamespace):
+    """
+    A :mod:`numpy` mimic for :class:`JAXArrayContext`.
+    """
+    def _get_fake_numpy_linalg_namespace(self):
+        return JAXFakeNumpyLinalgNamespace(self._array_context)
+
+    _jax_funcs = {"abs", "sin", "cos", "tan", "arcsin", "arccos",
+                  "arctan", "sinh", "cosh", "tanh", "exp", "log", "log10",
+                  "isnan", "sqrt", "exp", "conj"}
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/fake_numpy.py]: Yep, got rid of these. Thanks!

File: arraycontext/impl/jax/fake_numpy.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+
+
+class JAXFakeNumpyLinalgNamespace(BaseFakeNumpyLinalgNamespace):
+    # Everything is implemented in the base class for now.
+    pass
+
+
+class JAXFakeNumpyNamespace(BaseFakeNumpyNamespace):
+    """
+    A :mod:`numpy` mimic for :class:`JAXArrayContext`.
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/fake_numpy.py]: Thanks!

File: arraycontext/impl/jax/fake_numpy.py [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+from arraycontext.container.traversal import (
+        rec_multimap_array_container, rec_map_array_container,
+        rec_map_reduce_array_container,
+        )
+from arraycontext.container import NotAnArrayContainerError, serialize_container
+import numpy
+import jax.numpy as jnp
+
+
+class JAXFakeNumpyLinalgNamespace(BaseFakeNumpyLinalgNamespace):
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/fake_numpy.py]: Thanks!

File: arraycontext/impl/jax/fake_numpy.py [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+                return jnp.ravel(a, order=order)
+            elif order in "AK":
+                # flattening in a C-order
+                # memory layout is assumed to be "C"
+                return jnp.ravel(a, order="C")
+            else:
+                raise ValueError("`order` can be one of 'F', 'C', 'A' or 'K'. "
+                                 f"(got {order})")
+
+        return rec_map_array_container(_rec_ravel, a)
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply arraycontext/impl/jax/fake_numpy.py]: I see! 

Better to raise a `NotImplementedError` like `jax` here? It seems weird to silently change the order.

File: arraycontext/impl/jax/__init__.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    def from_numpy(self, array: Union[np.ndarray, _ScalarLike]):
+        import jax
+        return jax.device_put(array)
+
+    def to_numpy(self, array):
+        if np.isscalar(array):
+            return array
+
+        import jax
+        return jax.device_get(array)
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/__init__.py]: I think you're suggesting that we should always use `jax.device_get` and I agree with that. Implemented it. (Sorry, if I misunderstood)

[the remark this answers is not in the document]
Comment (alexfikl) [reply arraycontext/impl/jax/__init__.py]: I wasn't sure if it caused any inconsistency, but if it works, yeah, that would be my preference too!

File: arraycontext/impl/jax/fake_numpy.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (alexfikl) [reply arraycontext/impl/jax/fake_numpy.py]: Hm, yeah, fair enough.

[the remark this answers is not in the document]
Comment (inducer) [reply arraycontext/impl/jax/fake_numpy.py]: Maybe warn about this behavior?

**Review by inducer — APPROVED** [review]
L (almost) GTM! Two very minor follow-ups to @alexfikl's review, then this can land.

File: arraycontext/impl/pytato/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py]: Oops, thanks. Fixed for real now.

File: arraycontext/impl/jax/fake_numpy.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply arraycontext/impl/jax/fake_numpy.py]: Yep, warned. Thanks!

**Comment by kaushikcfd** [comment]
alexfikl, inducer: Thanks a lot for the suggestions!

**Pull request merged**